# Выпускной проект по SQL

## Описание проекта

Коронавирус застал мир врасплох, изменив привычный порядок вещей. В свободное время жители городов больше не выходят на улицу, не посещают кафе и торговые центры. Зато стало больше времени для книг. Это заметили стартаперы — и бросились создавать приложения для тех, кто любит читать.

Наша компания решила быть на волне и купила крупный сервис для чтения книг по подписке. 

**Цель исследования:** проанализировать базу данных, в которой содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные помогут сформулировать ценностное предложение для нового продукта.

## Описание данных

**Таблица books**
Содержит данные о книгах:
- book_id — идентификатор книги;
- author_id — идентификатор автора;
- title — название книги;
- num_pages — количество страниц;
- publication_date — дата публикации книги;
- publisher_id — идентификатор издателя.

**Таблица authors**
Содержит данные об авторах:
- author_id — идентификатор автора;
- author — имя автора.

**Таблица publishers**
Содержит данные об издательствах:
- publisher_id — идентификатор издательства;
- publisher — название издательства.

**Таблица ratings**
Содержит данные о пользовательских оценках книг:
- rating_id — идентификатор оценки;
- book_id — идентификатор книги;
- username — имя пользователя, оставившего оценку;
- rating — оценка книги.

**Таблица reviews**
Содержит данные о пользовательских обзорах:
- review_id — идентификатор обзора;
- book_id — идентификатор книги;
- username — имя автора обзора;
- text — текст обзора.

## Задачи

1. Посчитать, сколько книг вышло после 1 января 2000 года;
2. Для каждой книги посчитать количество обзоров и среднюю оценку;
3. Определить издательство, которое выпустило наибольшее число книг толще 50 страниц;
4. Определить автора с самой высокой средней оценкой книг;
5. Посчитать среднее количество обзоров от пользователей, которые поставили больше 50 оценок.

In [1]:
# импортируем библиотеки
import pandas as pd
from sqlalchemy import create_engine

In [2]:
# устанавливаем параметры
db_config = {'user': 'praktikum_student', # имя пользователя
'pwd': 'Sdf4$2;d-d30pp', # пароль
'host': 'rc1b-wcoijxj3yxfsf3fs.mdb.yandexcloud.net',
'port': 6432, # порт подключения
'db': 'data-analyst-final-project-db'} # название базы данных
connection_string = 'postgresql://{}:{}@{}:{}/{}'.format(db_config['user'],
 db_config['pwd'],
 db_config['host'],
 db_config['port'],
 db_config['db'])
# сохраняем коннектор
engine = create_engine(connection_string, connect_args={'sslmode':'require'}) 

In [3]:
# Напишем функцию для выполнения запросов.
def select (sql):
    return pd.io.sql.read_sql(sql, con=engine)

In [4]:
books = '''
        SELECT *
        FROM books
        LIMIT 5;
        '''
select (books)

book_id  author_id                                              title  \
0        1        546                                       'Salem's Lot   
1        2        465                 1 000 Places to See Before You Die   
2        3        407  13 Little Blue Envelopes (Little Blue Envelope...   
3        4         82  1491: New Revelations of the Americas Before C...   
4        5        125                                               1776   

   num_pages publication_date  publisher_id  
0        594       2005-11-01            93  
1        992       2003-05-22           336  
2        322       2010-12-21           135  
3        541       2006-10-10           309  
4        386       2006-07-04           268

Таблица **books** cодержит данные о книгах.

In [5]:
authors = '''
        SELECT *
        FROM authors
        LIMIT 5;
        '''
select (authors)

author_id                          author
0          1                      A.S. Byatt
1          2  Aesop/Laura Harris/Laura Gibbs
2          3                 Agatha Christie
3          4                   Alan Brennert
4          5        Alan Moore/David   Lloyd

Таблица **authors** cодержит данные об авторах.

In [6]:
publishers = '''
        SELECT *
        FROM publishers
        LIMIT 5;
        '''
select (publishers)

publisher_id                          publisher
0             1                                Ace
1             2                           Ace Book
2             3                          Ace Books
3             4                      Ace Hardcover
4             5  Addison Wesley Publishing Company

Таблица **publishers** cодержит данные об издательствах.

In [7]:
ratings = '''
        SELECT *
        FROM ratings
        LIMIT 5;
        '''
select (ratings)

rating_id  book_id       username  rating
0          1        1     ryanfranco       4
1          2        1  grantpatricia       2
2          3        1   brandtandrea       5
3          4        2       lorichen       3
4          5        2    mariokeller       2

Таблица **ratings** cодержит данные о пользовательских оценках книг.

In [8]:
reviews = '''
        SELECT *
        FROM reviews
        LIMIT 5;
        '''
select (reviews)

review_id  book_id       username  \
0          1        1   brandtandrea   
1          2        1     ryanfranco   
2          3        2       lorichen   
3          4        3  johnsonamanda   
4          5        3    scotttamara   

                                                text  
0  Mention society tell send professor analysis. ...  
1  Foot glass pretty audience hit themselves. Amo...  
2  Listen treat keep worry. Miss husband tax but ...  
3  Finally month interesting blue could nature cu...  
4  Nation purpose heavy give wait song will. List...

Таблица **reviews** cодержит данные о пользовательских обзорах на книги.

### Посчитаем, сколько книг вышло после 1 января 2000 года;

In [9]:
# Напишем функцию для выполнения запросов.
def select (sql):
    return pd.io.sql.read_sql(sql, con=engine)

In [10]:
sql = '''
        SELECT COUNT(book_id) books_count
        FROM books
        WHERE publication_date > '2000-01-01';
        '''
select (sql)

books_count
0          819

В базе данных содержится информация о 819 книгах, дата публикации которых после 1 января 2000 года.

### Для каждой книги посчитаем количество обзоров и среднюю оценку

In [11]:
sql = '''
        SELECT  b.title,
                COUNT(DISTINCT re.review_id) review_count,
                AVG(ra.rating) rating_avg
        FROM books b
        LEFT JOIN reviews re ON b.book_id = re.book_id
        JOIN ratings ra ON b.book_id = ra.book_id
        GROUP BY b.book_id
        ORDER BY 3 DESC;
        '''
select (sql)

title  review_count  \
0        Arrows of the Queen (Heralds of Valdemar  #1)             2   
1    The Walking Dead  Book One (The Walking Dead #...             2   
2                                      Light in August             2   
3    Wherever You Go  There You Are: Mindfulness Me...             2   
4    Captivating: Unveiling the Mystery of a Woman'...             2   
..                                                 ...           ...   
995  The World Is Flat: A Brief History of the Twen...             3   
996                  His Excellency: George Washington             2   
997                                      Drowning Ruth             3   
998                                              Junky             2   
999                               Harvesting the Heart             2   

     rating_avg  
0          5.00  
1          5.00  
2          5.00  
3          5.00  
4          5.00  
..          ...  
995        2.25  
996        2.00  
997        2.00  
998        2.00  
999        1.50  

[1000 rows x 3 columns]

В базе данных имеются книги с максимальным средним рейтингом 5,0. Минимальным средним рейтингом - 1,5 обладает книга Harvesting the Heart.


In [12]:
sql = '''
        SELECT  b.title,
                COUNT(DISTINCT re.review_id) review_count,
                AVG(ra.rating) rating_avg
        FROM books b
        LEFT JOIN reviews re ON b.book_id = re.book_id
        JOIN ratings ra ON b.book_id = ra.book_id
        GROUP BY b.book_id
        ORDER BY 2 DESC;
        '''
select (sql)

title  review_count  \
0                              Twilight (Twilight  #1)             7   
1                                  Water for Elephants             6   
2                                     The Glass Castle             6   
3    Harry Potter and the Prisoner of Azkaban (Harr...             6   
4    The Curious Incident of the Dog in the Night-Time             6   
..                                                 ...           ...   
995    Anne Rice's The Vampire Lestat: A Graphic Novel             0   
996                            The Natural Way to Draw             0   
997   The Cat in the Hat and Other Dr. Seuss Favorites             0   
998                          Essential Tales and Poems             0   
999  Disney's Beauty and the Beast (A Little Golden...             0   

     rating_avg  
0      3.662500  
1      3.977273  
2      4.206897  
3      4.414634  
4      4.081081  
..          ...  
995    3.666667  
996    3.000000  
997    5.000000  
998    4.000000  
999    4.000000  

[1000 rows x 3 columns]

Если рассмотреть количество обзоров на книгу, то их не так уж много. Максимум - 7 обзоров - у книги Twilight (Сумерки).

### Определим издательство, которое выпустило наибольшее число книг толще 50 страниц

In [13]:
sql = '''
        SELECT p.publisher,
                COUNT(b.book_id)
        FROM publishers p
        JOIN books b ON p.publisher_id = b.publisher_id
        WHERE b.num_pages > 50
        GROUP BY 1
        ORDER BY 2 DESC
        LIMIT 1;
        '''
select (sql)

publisher  count
0  Penguin Books     42

Издательство, которое выпустило наибольшее число книг толще 50 страниц - это Penguin Books, выпустившее 42 книги.

### Определим автора с самой высокой средней оценкой книг (учитывая только книги с 50 и более оценками)

In [14]:
sql = '''
        SELECT books_50.author author,
                AVG(books_50.rating_avg) rating_avg
        FROM
            (SELECT a.author,
                    b.book_id,
                    AVG(r.rating) rating_avg,
                    COUNT(r.rating) rating_count
            FROM authors a
            INNER JOIN books b ON b.author_id = a.author_id
            INNER JOIN ratings r ON r.book_id = b.book_id
            GROUP BY 1, 2
            HAVING COUNT(r.rating) >= 50) books_50
        GROUP BY 1
        ORDER BY 2 DESC
        LIMIT 1     
        '''
select (sql)

author  rating_avg
0  J.K. Rowling/Mary GrandPré    4.283844

Самая высокая средняя оценка книг (4.28) у автора J.K.Rowling. 

Это неудивительно, так как книги о Гарри Поттере, автором которых она является, получили несколько наград и были проданы в количестве более 500 миллионов экземпляров. Они стали самой продаваемой серией книг в истории. (Материал из Википедии)

### Посчитаем среднее количество обзоров от пользователей, которые поставили больше 50 оценок

In [15]:
sql = ''' 
    SELECT avg(count) review_avg
    FROM
        (SELECT COUNT(review_id)
        FROM reviews r
        FULL OUTER JOIN (SELECT username, 
                        COUNT(rating_id) count 
                        FROM ratings 
                        GROUP BY 1) rating_count
        ON rating_count.username = r.username 
        WHERE count > 50
        GROUP BY r.username) as count;
'''
select (sql)

review_avg
0   24.333333

Пользователи, которые поставили более 50 оценок, в среднем составляют 24 обзора.

## Выводы

Целью исследования было проанализировать базу данных, в которой содержится информация о книгах, издательствах, авторах, а также пользовательские обзоры книг. Эти данные должны помочь сформулировать ценностное предложение для нового продукта.

Вот что удалось выявить в ходе исследования:
- В базе данных содержится информация о 819 книгах, дата публикации которых после 1 января 2000 года.
- В базе данных имеются книги с максимальным средним рейтингом 5,0. Минимальным средним рейтингом - 1,5 обладает книга Harvesting the Heart. Если рассмотреть количество обзоров на книгу, то их не так уж много. Максимум - 7 обзоров - у книги Twilight (Сумерки).
- Издательство, которое выпустило наибольшее число книг толще 50 страниц - это Penguin Books, выпустившее 42 книги.
- Самая высокая средняя оценка книг (4.28) у автора J.K.Rowling.
- Пользователи, которые поставили более 50 оценок, в среднем составляют 24 обзора.